neo/Consensus/ConsensusContext.cs

```diff
@@ -210,14 +210,44 @@ public ConsensusPayload MakePrepareRequest()
         {
             byte[] buffer = new byte[sizeof(ulong)];
             random.NextBytes(buffer);
+            Block.ConsensusData.Nonce = BitConverter.ToUInt64(buffer, 0);
+
             IEnumerable<Transaction> memoryPoolTransactions = Blockchain.Singleton.MemPool.GetSortedVerifiedTransactions();
             foreach (IPolicyPlugin plugin in Plugin.Policies)
                 memoryPoolTransactions = plugin.FilterForBlock(memoryPoolTransactions);
+
             List<Transaction> transactions = memoryPoolTransactions.ToList();
-            TransactionHashes = transactions.Select(p => p.Hash).ToArray();
-            Transactions = transactions.ToDictionary(p => p.Hash);
+
+            Transactions = new Dictionary<UInt256, Transaction>();
+            uint maxBlockSize = NativeContract.Policy.GetMaxBlockSize(Snapshot);
+            TransactionHashes = new UInt256[Math.Min(transactions.Count, NativeContract.Policy.GetMaxTransactionsPerBlock(Snapshot))];
+
+            // Prevent that block exceed the max size
+
+            Block.Transactions = new Transaction[0];
+            var fixedSize = Block.Size + IO.Helper.GetVarSize(TransactionHashes.Length); // ensure that the var size grows without exceed the max size
+
+            for (int x = 0, max = TransactionHashes.Length; x < max; x++)
+            {
+                var tx = transactions[x];
+
+                // Check if exceed
+                if (fixedSize + UInt256.Length + tx.Size > maxBlockSize) break;
+
+                TransactionHashes[x] = tx.Hash;
+                Transactions.Add(tx.Hash, tx);
+            }
+
+            // Truncate null values
+
+            if (TransactionHashes.Length > Transactions.Count)
+            {
+                Array.Resize(ref TransactionHashes, Transactions.Count);
```

Comment:
Complex loop but looks clear at the end...aehjauheuaea

It needs more comments, because otherwise all the time we are going to need to understand it again...jajajaja